File: reference-python/backend/planners/astar.py
"""A* — grid tabanlı yol arama."""

import heapq
import math
import itertools
import time
from typing import Sequence

from backend.planners.base import Obstacle, PlanResult, Planner, Point

_NEIGHBOR_OFFSETS = [
    offset for offset in itertools.product((-1, 0, 1), repeat=3) if offset != (0, 0, 0)
]


class AStarPlanner(Planner):
    """Sabit çözünürlüklü 3B grid üzerinde A* ile yol arar."""

    name = "astar"

    def __init__(
        self,
        collision_checker,
        resolution: float = 0.05,
        padding: float = 0.2,
        max_expansions: int = 50000,
    ) -> None:
        super().__init__(collision_checker)
        self.resolution = resolution
        self.padding = padding
        self.max_expansions = max_expansions

    def _to_cell(self, point: Point) -> tuple:
        return tuple(round(coord / self.resolution) for coord in point)

    def _to_point(self, cell: tuple) -> Point:
        return tuple(index * self.resolution for index in cell)

    def _heuristic(self, cell: tuple, goal_cell: tuple) -> float:
        return sum((a - b) ** 2 for a, b in zip(cell, goal_cell)) ** 0.5 * self.resolution

    def _segment_free(self, a: Point, b: Point) -> bool:
        """İki nokta arasındaki doğru parçası serbest mi.

        Yalnızca hücre merkezlerini test etmek yetmiyor: ızgara
        çözünürlüğünden ince bir engel iki merkez arasına sığdığında
        görünmez oluyor ve yol engelin içinden geçiyordu.
        """
        dist = sum((x - y) ** 2 for x, y in zip(a, b)) ** 0.5
        steps = max(1, math.ceil(dist / (self.resolution / 2)))
        for step in range(1, steps + 1):
            t = step / steps
            point = tuple(a[i] + (b[i] - a[i]) * t for i in range(3))
            if not self.collision_checker(point):
                return False
        return True

    def _diagonal_allowed(self, cell: tuple, offset: tuple) -> bool:
        """Çapraz hamlenin geçtiği ara hücreler de serbest mi.

        Yalnızca hedef hücreyi kontrol etmek yetmiyor: iki dolu hücre
        köşegen olarak komşuysa, aradaki boşluktan geçen bir hamle
        geometrik olarak engelin içinden geçer. Her eksen bileşenini tek
        başına uygulayıp o ara hücrenin de serbest olmasını şart koşuyoruz.
        """
        for axis, delta in enumerate(offset):
            if delta == 0:
                continue
            step = [0, 0, 0]
            step[axis] = delta
            between = tuple(c + s for c, s in zip(cell, step))
            if not self.collision_checker(self._to_point(between)):
                return False
        return True

    def plan(self, start: Point, goal: Point, obstacles: Sequence[Obstacle]) -> PlanResult:
        started_at = time.perf_counter()

        if not self.collision_checker(start) or not self.collision_checker(goal):
            return PlanResult(
                success=False, elapsed_seconds=time.perf_counter() - started_at, algorithm=self.name
            )

        start_cell = self._to_cell(start)
        goal_cell = self._to_cell(goal)
        padding_cells = max(1, round(self.padding / self.resolution))
        bounds = [
            (min(s, g) - padding_cells, max(s, g) + padding_cells)
            for s, g in zip(start_cell, goal_cell)
        ]

        def in_bounds(cell: tuple) -> bool:
            return all(low <= c <= high for c, (low, high) in zip(cell, bounds))

        open_heap = [(self._heuristic(start_cell, goal_cell), 0.0, start_cell)]
        came_from: dict = {}
        cost_so_far = {start_cell: 0.0}
        visited = set()
        nodes_expanded = 0

        while open_heap and nodes_expanded < self.max_expansions:
            _, cost, current = heapq.heappop(open_heap)
            if current in visited:
                continue
            visited.add(current)
            nodes_expanded += 1

            if current == goal_cell:
                path_cells = [current]
                while current in came_from:
                    current = came_from[current]
                    path_cells.append(current)
                path_cells.reverse()
                # Yol, ilk ve son hücre merkezi yerine tam start/goal ile
                # kapanıyor; bu iki bağlantı segmenti ızgarada yer almadığı
                # için ayrıca kontrol edilmeli.
                path = [start] + [self._to_point(c) for c in path_cells[1:-1]] + [goal]
                if any(
                    not self._segment_free(path[i - 1], path[i])
                    for i in range(1, len(path))
                ):
                    return PlanResult(
                        success=False,
                        path=[],
                        elapsed_seconds=time.perf_counter() - started_at,
                        nodes_expanded=nodes_expanded,
                        algorithm=self.name,
                    )
                return PlanResult(
                    success=True,
                    path=path,
                    elapsed_seconds=time.perf_counter() - started_at,
                    nodes_expanded=nodes_expanded,
                    algorithm=self.name,
                )

            for offset in _NEIGHBOR_OFFSETS:
                neighbor = tuple(c + o for c, o in zip(current, offset))
                if neighbor in visited or not in_bounds(neighbor):
                    continue
                if not self.collision_checker(self._to_point(neighbor)):
                    continue
                # Köşe kesme (corner cutting) engeli: çapraz bir hamle, ancak
                # her bir eksen bileşeni tek başına da serbestse yapılabilir.
                # Bu olmadan robot iki dolu hücrenin arasından "sızarak"
                # engelin köşesini kesiyordu.
                if not self._diagonal_allowed(current, offset):
                    continue
                if not self._segment_free(self._to_point(current), self._to_point(neighbor)):
                    continue
                step_cost = sum(o ** 2 for o in offset) ** 0.5 * self.resolution
                new_cost = cost_so_far[current] + step_cost
                if neighbor not in cost_so_far or new_cost < cost_so_far[neighbor]:
                    cost_so_far[neighbor] = new_cost
                    priority = new_cost + self._heuristic(neighbor, goal_cell)
                    heapq.heappush(open_heap, (priority, new_cost, neighbor))
                    came_from[neighbor] = current

        return PlanResult(
            success=False,
            path=[],
            elapsed_seconds=time.perf_counter() - started_at,
            nodes_expanded=nodes_expanded,
            algorithm=self.name,
        )

File: lib/robotics/planners/astar.ts
import type { Vec3 } from "../transform";
import type { CollisionChecker, PlanResult, Planner } from "./base";

/** Grid hücresi: [x, y, z] tamsayı indeksleri. */
type Cell = readonly [number, number, number];

function cellKey(cell: Cell): string {
  return `${cell[0]},${cell[1]},${cell[2]}`;
}

function cellCompare(a: Cell, b: Cell): number {
  return a[0] - b[0] || a[1] - b[1] || a[2] - b[2];
}

const NEIGHBOR_OFFSETS: Cell[] = (() => {
  const offsets: Cell[] = [];
  for (const dx of [-1, 0, 1]) {
    for (const dy of [-1, 0, 1]) {
      for (const dz of [-1, 0, 1]) {
        if (dx === 0 && dy === 0 && dz === 0) continue;
        offsets.push([dx, dy, dz]);
      }
    }
  }
  return offsets;
})();

interface HeapEntry {
  priority: number;
  cost: number;
  cell: Cell;
}

/**
 * Basit binary min-heap. Python'daki heapq'nin (priority, cost, cell) tuple
 * sıralamasıyla birebir eşleşsin diye üç seviyeli karşılaştırma yapar —
 * bu, eşit öncelikli komşuların çok olduğu simetrik ızgaralarda Python
 * fixture'ıyla aynı düğüm genişletme sırasını (ve sayısını) üretmek için
 * gerekli.
 */
class MinHeap {
  private items: HeapEntry[] = [];

  get size() {
    return this.items.length;
  }

  push(entry: HeapEntry) {
    this.items.push(entry);
    this.bubbleUp(this.items.length - 1);
  }

  pop(): HeapEntry | undefined {
    if (this.items.length === 0) return undefined;
    const top = this.items[0];
    const last = this.items.pop()!;
    if (this.items.length > 0) {
      this.items[0] = last;
      this.bubbleDown(0);
    }
    return top;
  }

  private less(a: HeapEntry, b: HeapEntry): boolean {
    if (a.priority !== b.priority) return a.priority < b.priority;
    if (a.cost !== b.cost) return a.cost < b.cost;
    return cellCompare(a.cell, b.cell) < 0;
  }

  private bubbleUp(index: number) {
    let i = index;
    while (i > 0) {
      const parent = (i - 1) >> 1;
      if (!this.less(this.items[i], this.items[parent])) break;
      [this.items[i], this.items[parent]] = [this.items[parent], this.items[i]];
      i = parent;
    }
  }

  private bubbleDown(index: number) {
    let i = index;
    const n = this.items.length;
    for (;;) {
      const left = i * 2 + 1;
      const right = i * 2 + 2;
      let smallest = i;
      if (left < n && this.less(this.items[left], this.items[smallest])) smallest = left;
      if (right < n && this.less(this.items[right], this.items[smallest])) smallest = right;
      if (smallest === i) break;
      [this.items[i], this.items[smallest]] = [this.items[smallest], this.items[i]];
      i = smallest;
    }
  }
}

export interface AStarOptions {
  /** Izgara hücre boyutu (metre). */
  resolution?: number;
  /** Arama sınırlarını start/goal kutusunun dışına genişletme payı (metre). */
  padding?: number;
  maxExpansions?: number;
  /**
   * Düzlemsel arama: z ekseninde hiç hareket edilmez.
   *
   * 2B sahnelerde (PlannerRace üstten görünüm) planlayıcı 3B aradığı için
   * yol, kullanıcının göremediği z ekseninden dolaşıp engeli "delmiş" gibi
   * görünüyordu. Bkz. RrtOptions.planar.
   */
  planar?: boolean;
}

/**
 * A* — sabit çözünürlüklü 3B grid üzerinde arama.
 * reference-python/backend/planners/astar.py'nin TS portu (bkz. docs/02).
 */
export class AStarPlanner implements Planner {
  name = "astar";
  private resolution: number;
  private padding: number;
  private maxExpansions: number;
  private planar: boolean;

  constructor(options: AStarOptions = {}) {
    this.resolution = options.resolution ?? 0.05;
    this.padding = options.padding ?? 0.2;
    this.maxExpansions = options.maxExpansions ?? 50000;
    this.planar = options.planar ?? false;
  }

  private toCell(point: Vec3): Cell {
    return [
      Math.round(point.x / this.resolution),
      Math.round(point.y / this.resolution),
      Math.round(point.z / this.resolution),
    ];
  }

  private toPoint(cell: Cell): Vec3 {
    return { x: cell[0] * this.resolution, y: cell[1] * this.resolution, z: cell[2] * this.resolution };
  }

  /**
   * Çapraz hamlenin geçtiği ara hücreler de serbest mi.
   *
   * Yalnızca hedef hücreyi kontrol etmek yetmez: iki dolu hücre köşegen
   * komşuysa, aradaki boşluktan geçen hamle geometrik olarak engelin
   * içinden geçer (corner cutting). Her eksen bileşenini tek başına
   * uygulayıp o ara hücrenin de serbest olmasını şart koşuyoruz.
   * reference-python/backend/planners/astar.py `_diagonal_allowed` ile aynı.
   */
  private diagonalAllowed(cell: Cell, offset: Cell, isFree: CollisionChecker): boolean {
    for (let axis = 0; axis < 3; axis++) {
      if (offset[axis] === 0) continue;
      const between: Cell = [
        cell[0] + (axis === 0 ? offset[0] : 0),
        cell[1] + (axis === 1 ? offset[1] : 0),
        cell[2] + (axis === 2 ? offset[2] : 0),
      ];
      if (!isFree(this.toPoint(between))) return false;
    }
    return true;
  }

  /**
   * İki nokta arasındaki doğru parçası serbest mi.
   *
   * A* yalnızca hücre MERKEZLERİNİ test ediyordu; ızgara çözünürlüğünden
   * ince bir engel iki merkez arasına sığdığında görünmez oluyor ve yol
   * engelin içinden geçiyordu. Kenarları ve start/goal bağlantılarını
   * çözünürlüğün yarısı adımla örnekliyoruz.
   */
  private segmentFree(a: Vec3, b: Vec3, isFree: CollisionChecker): boolean {
    const dist = Math.hypot(b.x - a.x, b.y - a.y, b.z - a.z);
    const steps = Math.max(1, Math.ceil(dist / (this.resolution / 2)));
    for (let step = 1; step <= steps; step++) {
      const t = step / steps;
      const point: Vec3 = {
        x: a.x + (b.x - a.x) * t,
        y: a.y + (b.y - a.y) * t,
        z: a.z + (b.z - a.z) * t,
      };
      if (!isFree(point)) return false;
    }
    return true;
  }

  private heuristic(cell: Cell, goalCell: Cell): number {
    const dx = cell[0] - goalCell[0];
    const dy = cell[1] - goalCell[1];
    const dz = cell[2] - goalCell[2];
    return Math.sqrt(dx * dx + dy * dy + dz * dz) * this.resolution;
  }

  plan(start: Vec3, goal: Vec3, isFree: CollisionChecker): PlanResult {
    const startedAt = Date.now();

    if (!isFree(start) || !isFree(goal)) {
      return { success: false, path: [], elapsedMs: Date.now() - startedAt, nodesExpanded: 0, algorithm: this.name };
    }

    const startCell = this.toCell(start);
    const goalCell = this.toCell(goal);
    const goalKey = cellKey(goalCell);
    const paddingCells = Math.max(1, Math.round(this.padding / this.resolution));
    const bounds = [0, 1, 2].map((axis) => {
      const low = Math.min(startCell[axis], goalCell[axis]) - paddingCells;
      const high = Math.max(startCell[axis], goalCell[axis]) + paddingCells;
      return [low, high] as const;
    });
    const inBounds = (cell: Cell) => cell.every((c, axis) => c >= bounds[axis][0] && c <= bounds[axis][1]);

    const open = new MinHeap();
    open.push({ priority: this.heuristic(startCell, goalCell), cost: 0, cell: startCell });
    const cameFrom = new Map<string, Cell>();
    const costSoFar = new Map<string, number>([[cellKey(startCell), 0]]);
    const visited = new Set<string>();
    let nodesExpanded = 0;

    while (open.size > 0 && nodesExpanded < this.maxExpansions) {
      const entry = open.pop()!;
      const currentKey = cellKey(entry.cell);
      if (visited.has(currentKey)) continue;
      visited.add(currentKey);
      nodesExpanded++;

      if (currentKey === goalKey) {
        const cells: Cell[] = [entry.cell];
        let cursor = currentKey;
        while (cameFrom.has(cursor)) {
          const prev = cameFrom.get(cursor)!;
          cells.push(prev);
          cursor = cellKey(prev);
        }
        cells.reverse();
        // Yol, ilk ve son hücre merkezi yerine tam start/goal ile kapanıyor.
        // Bu iki bağlantı segmenti ızgarada hiç yer almadığı için ayrıca
        // kontrol edilmeli — aksi hâlde aralarında duran bir engel
        // görülmeden "başarılı" bir yol dönebilir.
        const path: Vec3[] = [start, ...cells.slice(1, -1).map((c) => this.toPoint(c)), goal];
        for (let i = 1; i < path.length; i++) {
          if (!this.segmentFree(path[i - 1], path[i], isFree)) {
            return {
              success: false,
              path: [],
              elapsedMs: Date.now() - startedAt,
              nodesExpanded,
              algorithm: this.name,
            };
          }
        }
        return { success: true, path, elapsedMs: Date.now() - startedAt, nodesExpanded, algorithm: this.name };
      }

      const offsets = this.planar
        ? NEIGHBOR_OFFSETS.filter((offset) => offset[2] === 0)
        : NEIGHBOR_OFFSETS;

      for (const offset of offsets) {
        const neighbor: Cell = [
          entry.cell[0] + offset[0],
          entry.cell[1] + offset[1],
          entry.cell[2] + offset[2],
        ];
        const neighborKey = cellKey(neighbor);
        if (visited.has(neighborKey) || !inBounds(neighbor)) continue;
        if (!isFree(this.toPoint(neighbor))) continue;
        if (!this.diagonalAllowed(entry.cell, offset, isFree)) continue;
        if (!this.segmentFree(this.toPoint(entry.cell), this.toPoint(neighbor), isFree)) continue;

        const stepCost = Math.sqrt(offset[0] ** 2 + offset[1] ** 2 + offset[2] ** 2) * this.resolution;
        const newCost = (costSoFar.get(currentKey) ?? 0) + stepCost;
        const existingCost = costSoFar.get(neighborKey);
        if (existingCost === undefined || newCost < existingCost) {
          costSoFar.set(neighborKey, newCost);
          open.push({ priority: newCost + this.heuristic(neighbor, goalCell), cost: newCost, cell: neighbor });
          cameFrom.set(neighborKey, entry.cell);
        }
      }
    }

    return { success: false, path: [], elapsedMs: Date.now() - startedAt, nodesExpanded, algorithm: this.name };
  }
}
